pkg/scheduler/scheduling_node.go

```diff
@@ -27,67 +27,158 @@ import (
 )
 
 type SchedulingNode struct {
-	NodeInfo *cache.NodeInfo
 	NodeId   string
 
-	// Resource which is allocating (in addition to confirmed, allocated)
-	AllocatingResource      *resources.Resource
-	PreemptingResource      *resources.Resource
-	CachedAvailableResource *resources.Resource
+	// Private info
+	nodeInfo           *cache.NodeInfo
+	allocatingResource *resources.Resource // resources being allocated
+	preemptingResource *resources.Resource // resources considered for preemption
+	cachedAvailable    *resources.Resource // calculated available resources
+	availableDirty     bool                // is the calculated available resource up to date?
 
 	lock sync.RWMutex
 }
 
 func NewSchedulingNode(info *cache.NodeInfo) *SchedulingNode {
+	// save guard against panic
+	if info == nil {
+		return nil
+	}
 	return &SchedulingNode{
-		NodeInfo:                info,
-		NodeId:                  info.NodeId,
-		AllocatingResource:      resources.NewResource(),
-		PreemptingResource:      resources.NewResource(),
-		CachedAvailableResource: info.GetAvailableResource(),
+		nodeInfo:           info,
+		NodeId:             info.NodeId,
+		allocatingResource: resources.NewResource(),
+		preemptingResource: resources.NewResource(),
+		availableDirty:     true,
 	}
 }
 
-func (m *SchedulingNode) CheckAndAllocateResource(delta *resources.Resource, preemptionPhase bool) bool {
-	m.lock.Lock()
-	m.lock.Unlock()
-	newAllocating := resources.Add(delta, m.AllocatingResource)
+// Get the allocated resource on this node.
+// These resources are just the confirmed allocations (tracked in the cache node).
+// This does not lock the cache node as it will take its own lock.
+func (sn *SchedulingNode) GetAllocatedResource() *resources.Resource {
+	return sn.nodeInfo.GetAllocatedResource()
+}
 
-	avail := m.CachedAvailableResource
-	if preemptionPhase {
-		avail = resources.Add(avail, m.PreemptingResource)
+// Get the available resource on this node.
+// These resources are confirmed allocations (tracked in the cache node) minus the resources
+// currently being allocated but not confirmed in the cache.
+// This does not lock the cache node as it will take its own lock.
+func (sn *SchedulingNode) getAvailableResource() *resources.Resource {
+    sn.lock.Lock()
+    defer sn.lock.Unlock()
+    if sn.availableDirty {
+		sn.cachedAvailable = resources.Sub(sn.nodeInfo.GetAvailableResource(), sn.allocatingResource)
+		sn.availableDirty = false
 	}
+	return sn.cachedAvailable
+}
+
+// Get the resource tagged for allocation on this node.
+// These resources are part of unconfirmed allocations.
+func (sn *SchedulingNode) getAllocatingResource() *resources.Resource {
+	sn.lock.RLock()
+	defer sn.lock.RUnlock()
+
+	return sn.allocatingResource
+}
+
+// Update the number of resource proposed for allocation on this node
+func (sn *SchedulingNode) incAllocatingResource(proposed *resources.Resource) {
+	sn.lock.Lock()
```

Comment:
NIT. extra space

Reply:
fixed removed spaces and replaced with tab